src/patterns.jl

```diff
@@ -0,0 +1,63 @@
+
+"""
+    Pattern
+
+A pattern is just a wrapper around a `Tuple{Vararg{Symbol}}` which enables searching for
+matching components and dimension paths in a Dataset.
+Special symbols `:_` and `:__` are used as wildcards, similar to `*` and `**` in glob
+pattern matching.
+"""
+struct Pattern
+    segments::Tuple{Vararg{Symbol}}
+end
+
+Pattern(segments::Symbol...) = Pattern(segments)
+Base.convert(::Type{Pattern}, x::Tuple{Vararg{Symbol}}) = Pattern(x)
+
+# Primary functionality is to support identify tuples that match the constraint
+# NOTE: I'm not entirely sure if we should overload `in` for this algorithm
+function Base.in(item::Tuple{Vararg{Symbol}}, pattern::Pattern)
+    pat = pattern.segments
+    i = 1   # item  index
+    j = 1   # contraint index
+    n = length(item)
+    m = length(pat)
+
+    # If our constrain has more segments than the item then it isn't going to match
+    m <= n || return false
+
+    result = true
+    while result && i <= n && j <= m
+        if pat[j] === item[i] || pat[j] === :_
+            i += 1
+            j += 1
+        elseif pat[j] === :__
+            # Lookahead if we aren't at the end
+            if j < m
+                _next = pat[j+1]
+
+                # If the next value matches the current item then just jump to the indices after both
+                if _next === item[i]
+                    i += 1
+                    j += 2
+                # Error if have multiple multiple wildcards in a row
```

Comment:
```suggestion
                # Error if we have mixed wildcards in a row
```

as I understand `(:_, :_, ...)` is ok but `(:__, :_, ...)` or `(:_, :__, ...)` isn't?

Reply:
Yeah, specifically, if we have `:_` directly after ` :__` then we don't know how many steps to skip. I suppose we could handle this by always iterating to the next non-wildcard item. The result would be that `(:__, :_, ...)` would essentially reduce to `(:__, ...)`.

Reply:
Alright, I've added a little logic to simplify the patterns at construction time. This will remove any unnecessary `:_`/`:__` patterns.